alacritty/src/renderer/rects.rs

```diff
@@ -190,3 +195,208 @@ impl RenderLines {
+            gl::VertexAttribPointer(
+                0,                            // Attribute location.
+                2,                            // Component count: 2. x and y.
+                gl::FLOAT,                    // Attribute type: f32
+                gl::FALSE,                    // Not normalized.
+                size_of::<Vertex>() as i32,   // Stride.
+                attribute_offset as *const _, // Offset of Vertex::x is zero.
+            );
+            gl::EnableVertexAttribArray(0);
+            attribute_offset += size_of::<f32>() * 2;
+
+            // Color.
+            gl::VertexAttribPointer(
+                1,                            // Attribute location.
+                4,                            // Component count: 4. r, g, b, a.
+                gl::UNSIGNED_BYTE,            // Attribute type: u8.
+                gl::TRUE,                     // Normalized: 0..255 is mapped to 0..1.
+                size_of::<Vertex>() as i32,   // Stride.
+                attribute_offset as *const _, // Offset of Vertex::color.
+            );
+            gl::EnableVertexAttribArray(1);
+
+            // Reset buffer bindings.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self { vao, vbo, program, vertices: Vec::new() })
+    }
+
+    pub fn draw(&mut self, size_info: &SizeInfo, rects: Vec<RenderRect>) {
+        unsafe {
+            // Bind VAO to enable vertex attribute slots.
+            gl::BindVertexArray(self.vao);
+
+            // Bind VBO only once for buffer data upload only.
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(self.program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        // Build rect vertices vector.
+        self.vertices.clear();
+        for rect in &rects {
+            self.add_rect(center_x, center_y, rect);
+        }
+
+        unsafe {
+            // Upload accumulated vertices.
+            gl::BufferData(
+                gl::ARRAY_BUFFER,
+                (self.vertices.len() * std::mem::size_of::<Vertex>()) as isize,
+                self.vertices.as_ptr() as *const _,
+                gl::STREAM_DRAW,
+            );
+
+            // Draw all vertices as list of triangles.
+            gl::DrawArrays(gl::TRIANGLES, 0, self.vertices.len() as i32);
+
+            // Disable program.
+            gl::UseProgram(0);
+
+            // Reset buffer bindings to nothing.
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+            gl::BindVertexArray(0);
+        }
+    }
+
+    fn add_rect(&mut self, center_x: f32, center_y: f32, rect: &RenderRect) {
+        // Calculate rectangle position.
+        let x = (rect.x - center_x) / center_x;
+        let y = -(rect.y - center_y) / center_y;
+        let width = rect.width / center_x;
+        let height = rect.height / center_y;
```

Comment:
As someone unfamiliar with OpenGL the purpose of these calculations isn't immediately obvious to me. Based on some rough guesses I'd assume you're constructing a normalized rectangle ranging with values from -1 to +1?

If that is the case, I think it would make sense to just state that nice and shortly, if it isn't please let me know what it does.

My suggestion would be something like this:

```
    // Calculate the normalized rectangle positions. These values range from -1 to +1.
```

Reply:
Just to clarify, this logic is straight from the old code. And it was without docs even there, so that's why I haven't mentioned during review.

Reply:
Oh, I missed that. The point still stands though so I think with my suggestion that would be a trivial improvement.